### PR TITLE
migrate to erc721a

### PR DESCRIPTION
**Pull request by sanderpick**
Many things done in this PR:

- Migrates contract to https://chiru-labs.github.io/ERC721A/#/upgradeable
- Reworked file layout and inheritance (modeled after ^)
- Added a complete test suite. It's now 100% coverage, including controller logic and proxy upgrading
- Reworked deployment and upgrade flow to inherit from `hardhat.config.ts`. Proxy management and baseURIs for different networks handled there. No editing of files needed. See comments.
- Added docstrings
- Updated prettier and eslint flows. The current recommendation from prettier is to not run prettier ad an eslint rule. This is pretty nice since you don't get tons of red flags when you're writing code. You just run `npm run format` to fix all format and eslint problems. Use prettier for formatting, eslint for code problems. Bonus, setup vscode to format on save -> back to Go-like heaven : ) I can share my `settings.json` if anyone is interested.
- Added lint step to CI. Bad format or code problems will fail the actions.
- Updated dependencies.

Changes to contract logic:
- The notion of what "paused" means was unclear. It was largely unused, but we included those extensions. In this PR, I have adopted "pause" to mean the halting of `createTable`, `runSQL`, and `setController`, but did not touch token transfers and other functions like burning.
- The access control seemed overkill. Here I'm just using `Ownable`.
- The `TransferTable` event had to change due to erc721a difference. I'll open a PR for `go-tableland` to fix that and update ABI + bytecode.

Screen Shot 2022-06-04 at 11 28 18 PM

### CODE REVIEW

**Comment by sanderpick** [comment]
Some progress:
- Rearranged files / controller layout
- Updated the main tests

TODO:
- Add some more tests
- Update proxy tests

File: .prettierrc.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
+  "semi": false,
```
Comment (sanderpick) [inline .prettierrc.json:2]: nothing worse than a semi colon!

File: README.md [hunk 2]
```diff
@@ -17,64 +17,67 @@
 - [Contributing](#contributing)
 - [License](#license)
 
-# Deployments
```
Comment (sanderpick) [inline README.md:20]: existing proxies are managed in hardhat config

File: contracts/test/TestTablelandTablesUpgrade.sol [hunk 3]
```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: MIT
```
Comment (sanderpick) [inline contracts/test/TestTablelandTablesUpgrade.sol:1]: used to test storage upgrades and controller compatibility

File: contracts/utils/ERC721AQueryablePolicies.sol [hunk 4]
```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import "erc721a/contracts/extensions/ERC721AQueryable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/StringsUpgradeable.sol";
+
+/**
+ * @dev Library containing {ERC721AQueryable}-related helper methods for writing {ITablelandController.Policy}s.
+ */
+library ERC721AQueryablePolicies {
```
Comment (sanderpick) [inline contracts/utils/ERC721AQueryablePolicies.sol:10]: these helper libs are split up so that people only need to compile the code that they need

File: hardhat.config.ts [hunk 5]
```diff
@@ -1,51 +1,106 @@
...
+dotenv.config()
 
 const config: HardhatUserConfig = {
   solidity: {
     version: "0.8.4",
     settings: {
       optimizer: {
         enabled: true,
-        runs: 2,
+        runs: 200,
```
Comment (sanderpick) [inline hardhat.config.ts:21]: bump to default value

File: hardhat.config.ts [hunk 6]
```diff
@@ -1,51 +1,106 @@
...
+    only: [],
+  },
+  gasReporter: {
+    enabled: process.env.REPORT_GAS !== undefined,
+    currency: "USD",
+  },
+  etherscan: {
+    apiKey: process.env.ETHERSCAN_API_KEY || "",
+  },
   networks: {
```
Comment (sanderpick) [inline hardhat.config.ts:39]: one network per tableland <-> chain setup

File: hardhat.config.ts [hunk 7]
```diff
@@ -55,13 +110,80 @@ const config: HardhatUserConfig = {
       },
     },
   },
-  gasReporter: {
-    enabled: process.env.REPORT_GAS !== undefined,
-    currency: "USD",
+  baseURIs: {
```
Comment (sanderpick) [inline hardhat.config.ts:113]: manage base URIs per network here

File: hardhat.config.ts [hunk 8]
```diff
@@ -55,13 +110,80 @@ const config: HardhatUserConfig = {
...
+    "optimism-kovan": "https://testnet.tableland.network/tables/",
+    "polygon-mumbai": "https://testnet.tableland.network/tables/",
+    // devnets
+    "ethereum-rinkeby-staging": "",
+    "optimism-kovan-staging": "https://staging.tableland.network/tables/",
+    localhost: "http://localhost:8080/tables/",
   },
-  etherscan: {
-    apiKey: process.env.ETHERSCAN_API_KEY || "",
+  proxies: {
```
Comment (sanderpick) [inline hardhat.config.ts:128]: manage existing proxies here

File: hardhat.config.ts [hunk 9]
```diff
@@ -55,13 +110,80 @@ const config: HardhatUserConfig = {
...
+
+declare module "hardhat/types/runtime" {
+  // eslint-disable-next-line no-unused-vars
+  interface HardhatRuntimeEnvironment {
+    baseURI: string
+    proxy: string
+  }
+}
+
+extendEnvironment((hre: HardhatRuntimeEnvironment) => {
```
Comment (sanderpick) [inline hardhat.config.ts:179]: extend the `hre` so that you can simply run `scripts/deploy.ts` or `scripts/upgrade.ts` and have the correct base URI and proxy injected.

File: package.json [hunk 10]
```diff
@@ -1,18 +1,22 @@
 {
   "name": "@tableland/eth",
-  "version": "0.1.0-dev.0",
-  "description": "On-chain ETH registry and client components for Tableland",
-  "main": "typechain/index.js",
-  "types": "typechain/index.d.ts",
+  "version": "0.0.2",
```
Comment (sanderpick) [inline package.json:3]: this is the current version

File: package.json [hunk 11]
```diff
@@ -1,18 +1,22 @@
...
     "build": "npx tsc -p ./tsconfig.build.json",
+    "up": "npm install && npx hardhat compile && npm run build && hardhat node",
     "test": "hardhat coverage",
-    "all": "npm install && npx hardhat compile && npm run build && npm test",
-    "up": "npm install && npx hardhat compile && npm run build && hardhat node"
+    "lint": "eslint '**/*.{js,ts}'",
+    "lint:fix": "npm run lint -- --fix",
+    "prettier": "prettier '**/*.{ts,json,sol,md}' --check",
+    "prettier:fix": "npm run prettier -- --write",
+    "format": "npm run prettier:fix && npm run lint:fix"
```
Comment (sanderpick) [inline package.json:19]: new commands to check lint and prettier for all files, and fix them too.

File: contracts/ITablelandController.sol [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * @dev Interface of a TablelandController compliant contract.
+ *
+ * This interface can be implemented to enabled advanced access control for a table.
+ * Call {ITablelandTables-setController} with the address of your implementation.
+ *
+ * See {test/TestTablelandController} for an example of token-gating table write-access.
+ */
+interface ITablelandController {
```
Comment (sanderpick) [inline contracts/ITablelandController.sol:12]: moved this to an interface instead of an abstract contract. this means that if the proxy is upgraded, controller contracts that refer to this interface will continue to serve policies to the tables contract. previously, this was not the case since the controller contract would refer to a contract referenced by the previous table implementation 😅

File: contracts/test/TestTablelandTablesUpgrade.sol (hunk 3 above)
Comment (carsonfarmer) [reply contracts/test/TestTablelandTablesUpgrade.sol:1]: excellent

File: hardhat.config.ts [hunk 13]
```diff
@@ -1,51 +1,107 @@
-import * as dotenv from "dotenv";
+import * as dotenv from "dotenv"
```
Comment (carsonfarmer) [inline hardhat.config.ts:1]: You know, I actually _like_ the semi-colons ;)

File: hardhat.config.ts (hunk 7 above)
Comment (carsonfarmer) [reply hardhat.config.ts:114]: this is really nice! didn't know about this "trick"

File: hardhat.config.ts (hunk 8 above)
Comment (carsonfarmer) [reply hardhat.config.ts:129]: nice! Reminder to us all: we'll have to update after we deploy

File: hardhat.config.ts (hunk 9 above)
Comment (carsonfarmer) [reply hardhat.config.ts:180]: cool

File: scripts/deploy.ts [hunk 14]
```diff
@@ -1,60 +1,66 @@
...
+  // Insert first row into health bot table
+  const runStatement = `insert into healthbot_${chainId}_${tableId} values (1);`
+  tx = await tables.runSQL(account.address, tableId, runStatement)
+  receipt = await tx.wait()
+  const [runEvent] = receipt.events ?? []
+  assert(runEvent.args!.statement === runStatement, "insert statement mismatch")
+  console.log("Healthbot table updated with:", runEvent.args?.statement)
 
-  console.log("Healthbot table updated with:", insertEvent.args?.statement);
+  // Warn that proxy address needs to be saved in config
```
Comment (carsonfarmer) [inline scripts/deploy.ts:57]: And there you go!

File: test/TablelandTables.ts [hunk 15]
```diff
@@ -0,0 +1,200 @@
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers"
+import chai from "chai"
+import chaiAsPromised from "chai-as-promised"
+import { BigNumber } from "ethers"
+// @ts-ignore-start
```
Comment (carsonfarmer) [inline test/TablelandTables.ts:5]: What's going on here? Why the ts-ignore?

File: test/TablelandTablesProxy.ts [hunk 16]
```diff
@@ -0,0 +1,191 @@
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers"
+import chai from "chai"
+import chaiAsPromised from "chai-as-promised"
+// @ts-ignore-start
+import { ethers, upgrades } from "hardhat"
```
Comment (carsonfarmer) [inline test/TablelandTablesProxy.ts:5]: this again? otherwise, very nice tests!

**Review by carsonfarmer — APPROVED** [review]
Look pretty good. I've left a few comments, some are just throwaway ones. The ts-igores in there always give me pause, can this be fixed?

File: .prettierrc.json (hunk 1 above)
Comment (sanderpick) [reply .prettierrc.json:2]: are you guys thumb-downing cause you agree, or you hate me now? haha

File: README.md (hunk 2 above)
Comment (sanderpick) [reply README.md:20]: i thought it would be annoying to have to update these in two places... i made a ref to the `proxies` entry in hardhat config in this file. we can def add back if you guys still want it

File: hardhat.config.ts (hunk 7 above)
Comment (sanderpick) [reply hardhat.config.ts:114]: 🙏 , me neither... had to dig for a solution

File: contracts/test/TestERC721AQueryable.sol [hunk 17]
```diff
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: MIT
```
[the remark this answers is not in the document]
Comment (sanderpick) [reply contracts/test/TestERC721AQueryable.sol:1]: Its used by the test controller

File: contracts/test/TestERC721Enumerable.sol [hunk 18]
```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
```
[the remark this answers is not in the document]
Comment (sanderpick) [reply contracts/test/TestERC721Enumerable.sol:2]: yep, used by the test controller

File: contracts/test/TestTablelandController.sol [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+            _bars,
+            "bar_id"
+        );
+
+        // Restrict updates to a single column
+        string[] memory updatableColumns = new string[](1);
+        updatableColumns[0] = "baz";
+
+        // Include a check on the incoming data
+        withChecks[0] = ""; // included to filter in Policies.joinClauses
```
[the remark this answers is not in the document]
Comment (sanderpick) [reply contracts/test/TestTablelandController.sol:42]: ah, should say "included to filter in Policies.joinClauses". making sure those lines are covered in that function.

File: hardhat.config.ts (hunk 13 above)
Comment (sanderpick) [reply hardhat.config.ts:1]: haha, really? i'll have to know why at some point

File: test/ITablelandController.ts [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+      await ethers.getContractFactory("TestAllowAllTablelandController")
+    ).deploy()) as TestAllowAllTablelandController
+    await allowAllController.deployed()
+
+    // Setup controller
+    await (await controller.setFoos(foos.address)).wait()
+    await (await controller.setBars(bars.address)).wait()
+  })
+
+  it("Should set controller for a table", async function () {
```
Comment (joewagner) [inline test/ITablelandController.ts:62]: If I want to "unset" the controller, do I set the Controller to the owner address?
Maybe we should add a test for "unsetting" the controller? By unsetting I mean get the table back to the same access control it would have before a controller is set.

File: .prettierrc.json (hunk 1 above)
Comment (joewagner) [reply .prettierrc.json:2]: LOL, I'm one of those Crockford troglodytes who likes semi-colons haha

File: hardhat.config.ts [hunk 21]
```diff
@@ -55,13 +111,80 @@ const config: HardhatUserConfig = {
       },
     },
   },
-  gasReporter: {
-    enabled: process.env.REPORT_GAS !== undefined,
-    currency: "USD",
+  baseURIs: {
+    // mainnets
+    ethereum: "https://tableland.network/tables/",
```
Comment (joewagner) [inline hardhat.config.ts:116]: Double checking, do we need to change these URIs to include the chainId?
ref: https://github.com/textileio/go-tableland/blob/2811fbfdefa7005bc4adc3b967920ddb18ee8317/cmd/api/main.go#L133

**Review by joewagner — APPROVED** [review]
Left a few comments...
TLDR; Everything looks great! It all works with the Chess game afaict. Maybe add one more test for unsetting a controller contract, and I think the baseURIs in the config might need to be changed (not totally sure on this though)

File: test/TablelandTables.ts (hunk 15 above)
Comment (sanderpick) [reply test/TablelandTables.ts:5]: it's odd... whenever i `import { ethers } from "hardhat"` i get this error

https://github.com/NomicFoundation/hardhat/issues/1627

but everything works fine, include using TS_NODE_FILES. i tried everything in that and other issues to get the red squiggle to go away.... maybe you have some magic tricks

Comment (sanderpick) [reply test/TablelandTables.ts:5]: oh, figured it out. ethers had to be in dev dependencies. must be something that changed in recent version

File: test/ITablelandController.ts (hunk 20 above)
Comment (sanderpick) [reply test/ITablelandController.ts:62]: oh good call! i'll add a test... you'd just set it to the zero address, but not sure what that looks like yet in TS

File: hardhat.config.ts (hunk 21 above)
Comment (sanderpick) [reply hardhat.config.ts:116]: damn, good catch! i think they do need to include chain id
